File: crates/korg-registry/tests/conformance.rs
//! korg-ledger@v1 cross-language conformance.
//!
//! The Python reference (korgex) froze the chain definition and a set of golden
//! vectors with FROZEN tip hashes (tests/conformance/conformance.json). This
//! test proves the Rust implementation reproduces those exact hashes — i.e. that
//! `korg-ledger@v1` is genuinely a multi-language standard, not a Python detail.
//! If Rust's canonicalization or hashing diverged by a single byte, the frozen
//! tip would not match and this test would fail.

use korg_registry::ledger_chain::{chain_hash, verify_chain, GENESIS_HASH};
use serde_json::Value;
use std::fs;
use std::path::PathBuf;

fn cdir() -> PathBuf {
    PathBuf::from(concat!(env!("CARGO_MANIFEST_DIR"), "/tests/conformance"))
}

fn read_jsonl(name: &str) -> Vec<Value> {
    fs::read_to_string(cdir().join(name))
        .unwrap()
        .lines()
        .filter(|l| !l.trim().is_empty())
        .map(|l| serde_json::from_str(l).unwrap())
        .collect()
}

#[test]
fn genesis_hash_is_64_zeros() {
    assert_eq!(GENESIS_HASH, "0".repeat(64));
}

#[test]
fn conformance_vectors_reproduce_the_frozen_oracle() {
    let manifest: Value =
        serde_json::from_str(&fs::read_to_string(cdir().join("conformance.json")).unwrap())
            .unwrap();
    assert_eq!(manifest["spec_version"], "korg-ledger@v1");
    let vectors = manifest["vectors"].as_array().unwrap();
    assert!(!vectors.is_empty(), "no conformance vectors");

    for v in vectors {
        let file = v["file"].as_str().unwrap();
        let events = read_jsonl(file);
        let key_owned = v["key"].as_str().map(|s| s.as_bytes().to_vec());
        let key = key_owned.as_deref();
        let errors = verify_chain(&events, key);

        match v["verify"].as_str().unwrap() {
            "intact" => {
                // verify_chain passing already proves Rust recomputes Python's
                // hashes (it compares chain_hash() to each stored entry_hash).
                assert!(errors.is_empty(), "{file}: expected intact, got {errors:?}");
                // and the explicit cross-impl anchor: Rust chain_hash == frozen tip.
                let tip = chain_hash(events.last().unwrap(), key);
                assert_eq!(
                    tip,
                    v["tip_entry_hash"].as_str().unwrap(),
                    "{file}: Rust chain_hash must reproduce the frozen tip"
                );
            }
            _ => {
                assert!(
                    !errors.is_empty(),
                    "{file}: expected tampered, verified clean"
                );
                let needle = v["error_contains"].as_str().unwrap();
                assert!(
                    errors.iter().any(|e| e.contains(needle)),
                    "{file}: errors {errors:?} missing {needle:?}"
                );
            }
        }
    }
}

#[test]
fn hmac_chain_fails_without_the_key() {
    // tamper-PROOF: a keyed chain verified with no key must fail.
    let events = read_jsonl("hmac-intact.jsonl");
    assert!(
        !verify_chain(&events, None).is_empty(),
        "keyed chain wrongly verified without the key"
    );
}

/// Non-BMP / surrogate-pair determinism oracle.
///
/// The events in `nonbmp-intact.jsonl` carry astral-plane code points —
/// 😀 (U+1F600), 💩 (U+1F4A9), 𐀀 (U+10000, the very first astral code point) —
/// alongside BMP CJK (中文). Python's `json.dumps(ensure_ascii=True)` escapes
/// each astral code point as a UTF-16 surrogate *pair* (`😀`, …),
/// which is exactly the `cp > 0xFFFF` branch in `write_json_string`. The frozen
/// tip `045d2d8…` was computed by the Python reference. If Rust's surrogate
/// arithmetic or non-ASCII escaping diverged by a single byte, this tip would
/// not reproduce — so this test is the empirical proof that the moat holds on
/// non-BMP text, not merely on ASCII.
#[test]
fn nonbmp_surrogate_chain_reproduces_python_frozen_tip() {
    const FROZEN_TIP: &str = "045d2d865e89c54340408b309c81d6b0f1367ebf0aa4504f1af66f7f1e4cb590";
    let events = read_jsonl("nonbmp-intact.jsonl");

    // The full chain must verify (each stored entry_hash recomputed by Rust).
    let errors = verify_chain(&events, None);
    assert!(
        errors.is_empty(),
        "non-BMP chain failed to verify: {errors:?}"
    );

    // Explicit cross-impl anchor: Rust's chain_hash of the tip event must equal
    // the Python-frozen tip, byte-for-byte through the surrogate-pair escaper.
    let tip = chain_hash(events.last().unwrap(), None);
    assert_eq!(
        tip, FROZEN_TIP,
        "Rust chain_hash diverged from Python on non-BMP text — moat-breaking \
         surrogate/escaping bug in write_canonical"
    );
}

File: spec/korg-ledger-v1/conformance.py
#!/usr/bin/env python3
"""
korg-ledger@v1 — standalone conformance verifier (the citable reference).

Dependency-free (Python stdlib only). Reproduces the spec's canonicalization +
hash-chain from scratch and checks every vector in conformance.json. Any
implementation in any language is conformant iff it reproduces the frozen
tip_entry_hash on each intact vector — this script is the executable oracle.

    python3 conformance.py        # exit 0 = the reference reproduces the vectors
"""
from __future__ import annotations

import hashlib
import hmac
import json
import os

HERE = os.path.dirname(os.path.abspath(__file__))
GENESIS = "0" * 64


def canonicalize(value) -> bytes:
    # JSON, keys sorted by code point, no whitespace, non-ASCII \uXXXX-escaped.
    return json.dumps(value, sort_keys=True, separators=(",", ":")).encode("ascii")


def chain_hash(event: dict, key: bytes | None = None) -> str:
    # entry_hash and the reserved Phase-2 event_sig are excluded from the preimage.
    preimage = {k: v for k, v in event.items() if k not in ("entry_hash", "event_sig")}
    data = canonicalize(preimage)
    if key is not None:
        return hmac.new(key, data, hashlib.sha256).hexdigest()
    return hashlib.sha256(data).hexdigest()


def verify_chain(events: list, key: bytes | None = None) -> list:
    errors, expected_prev = [], GENESIS
    for e in events:
        sid = e.get("seq_id")
        stored = e.get("entry_hash")
        if stored is None:
            errors.append(f"seq {sid}: missing entry_hash")
            expected_prev = None
            continue
        if e.get("prev_hash") != expected_prev:
            errors.append(f"seq {sid}: prev_hash breaks the chain")
        if chain_hash(e, key) != stored:
            errors.append(f"seq {sid}: entry_hash mismatch (content tampered)")
        expected_prev = stored
    return errors


def _read(name):
    with open(os.path.join(HERE, "vectors", name)) as f:
        return [json.loads(line) for line in f if line.strip()]


def run() -> int:
    manifest = json.load(open(os.path.join(HERE, "conformance.json")))
    assert manifest["spec_version"] == "korg-ledger@v1"
    failures = 0
    for v in manifest["vectors"]:
        events = _read(v["file"])
        key = v["key"].encode() if v.get("key") else None
        errors = verify_chain(events, key)
        ok, detail = True, ""
        if v["verify"] == "intact":
            if errors:
                ok, detail = False, f"expected intact, got {errors}"
            elif chain_hash(events[-1], key) != v["tip_entry_hash"]:
                ok, detail = False, "tip_entry_hash not reproduced"
        else:
            if not errors:
                ok, detail = False, "expected tampered, verified clean"
            elif not any(v["error_contains"] in e for e in errors):
                ok, detail = False, f"errors {errors} missing {v['error_contains']!r}"
        print(f"  [{'PASS' if ok else 'FAIL'}] {v['file']:<26} {v['verify']:<8} {detail}")
        failures += 0 if ok else 1
    print(f"\nkorg-ledger@v1 conformance: {'PASS' if not failures else f'{failures} FAILURE(S)'}")
    return 0 if not failures else 1


if __name__ == "__main__":
    raise SystemExit(run())
